rocr: Skip UnmapMemory/MapMemory operations in RemoveAccess for WSL

hongkzha-amd opened the pull request:
In WSL (Windows Subsystem for Linux) environments, the UnmapMemory and MapMemory operations by using drm fd during RemoveAccess can cause issues with memory management.

hongkzha-amd committed:
rocr: Add ProtectMemory API and use it in RemoveAccess

Replace munmap + mmap with mprotect when removing memory access.
This improves performance by 5-10x, ensures atomicity (no race
condition window), and prepares for WSL/DXG compatibility fixes.

Suggested-by: David Yat Sin <David.YatSin@amd.com>
Signed-off-by: Flora Cui <flora.cui@amd.com>
Signed-off-by: Horatio Zhang <Hongkun.Zhang@amd.com>

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/runtime.cpp:
```diff
@@ -3827,12 +3830,17 @@ hsa_status_t Runtime::MappedHandleAllowedAgent::RemoveAccess() {
-      if (!rocr::os::MapMemory(va, size, PermissionsToMemProt(perms), mappedHandle->drm_fd,
-                                reinterpret_cast<uint64_t>(mappedHandle->drm_cpu_addr))) {
-        return HSA_STATUS_ERROR;
+      else{
```
I think you will get an error on DXG on shutdown if the user application creates a VMM allocation and allows access to the CPU because it will hit the "else" at line 3842:

i,e a HIP application like this:
-------------------------
 hipMemGenericAllocationHandle_t handle;
 HIP_CHECK(hipMemCreate(&handle, size_aligned, &prop, 0));

  // 2. Reserve virtual address space
 hipDeviceptr_t ptrA;
 HIP_CHECK(hipMemAddressReserve(&ptrA, size_aligned, 0, 0, 0));

 // 3. Map the physical memory into the VA space
 HIP_CHECK(hipMemMap(ptrA, size_aligned, 0, handle, 0));

  // 4. Set GPU and GPU access
  hipMemAccessDesc accessDesc[2];
  accessDesc[0].location.type = hipMemLocationTypeDevice;
  accessDesc[0].location.id = device;
  accessDesc[0].flags = hipMemAccessFlagsProtReadWrite;

  **accessDesc[1].location.type = hipMemLocationTypeHost;
  accessDesc[1].flags = hipMemAccessFlagsProtReadWrite;**
  
  HIP_CHECK(hipMemSetAccess(ptrA, size_aligned, accessDesc, 2));

-------------------------

Looking at this code again, it looks like we should call mprotect instead of munmap + mmap.

Can you see if replacing these lines:

if (!rocr::os::UnmapMemory(va, size)) {
   return HSA_STATUS_ERROR;
}
if (!rocr::os::MapMemory(va, size, PermissionsToMemProt(perms), mappedHandle->drm_fd,
    reinterpret_cast<uint64_t>(mappedHandle->drm_cpu_addr))) {
    return HSA_STATUS_ERROR;
}

With a new os::ProtectMemory(..) would work for you?

if (!rocr::os::ProtectMemory(va, size, PermissionsToMemProt(perms)) {
    return HSA_STATUS_ERROR;
}


In os_linux.cpp:
bool ProtectMemory(void* va, size_t size, MemProt perms) {
     return ::mprotect(va, size, MemProtToOsProt(perms));
}

hongkzha-amd replied:
Thanks for the excellent suggestion! Using mprotect is indeed much cleaner and more efficient than munmap + mmap. I've updated the patch to use ProtectMemory as you recommended.

Regarding the WSL/DXG case - you're right that returning an error here isn't ideal for applications that request CPU access. Unfortunately, WSL has architectural restrictions on mapping GPU VRAM into CPU address space. If we were to use mprotect and allow CPU access, we'd run into some subtle issues:

1. CPU accesses would trigger page faults that allocate system RAM instead of mapping to GPU VRAM
2. This creates a silent data coherency issue - CPU and GPU would see different physical memory
3. The allocated system memory wouldn't be properly freed (mprotect doesn't release pages)

For now, I've kept the error return on WSL to fail fast and avoid the coherency/memory leak issues. If there's a way to make CPU access work correctly on WSL, I'd be happy to update the implementation.

dayatsin-amd commented:
Adding @manocharahul  to confirm whether this works on Windows.

manocharahul commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/runtime.cpp:
```diff
@@ -3827,12 +3830,17 @@ hsa_status_t Runtime::MappedHandleAllowedAgent::RemoveAccess() {
       hsa_access_permission_t perms = HSA_ACCESS_PERMISSION_NONE;
-      if (!rocr::os::UnmapMemory(va, size)) {
-        return HSA_STATUS_ERROR;
+      if (!core::Runtime::runtime_singleton_->thunkLoader()->IsDXG()) {
```
Using this check for IsDXG will fail VMM calls on windows. Is it possible to modify this check to work on windows?

manocharahul replied:
I had to remove IsDXG check from EnableAccess to make it work in windows
https://github.com/ROCm/rocm-systems/pull/2280

hongkzha-amd replied:
Thanks for your suggestion. I've added `#if defined(__linux__)` guards around the IsDXG checks to ensure they only apply on Linux. On Windows, the code will skip the IsDXG check entirely and proceed with the normal MapMemory/ProtectMemory path, so VMM calls should work as expected.

manocharahul commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/runtime.cpp:
```diff
@@ -3827,12 +3830,17 @@ hsa_status_t Runtime::MappedHandleAllowedAgent::RemoveAccess() {
-      if (!rocr::os::UnmapMemory(va, size)) {
-        return HSA_STATUS_ERROR;
+      if (!core::Runtime::runtime_singleton_->thunkLoader()->IsDXG()) {
+        if (!rocr::os::UnmapMemory(va, size)) {
```
on windows we have already created abstraction for map/unmap calls these calls work fine.

hongkzha-amd replied:
Thank you for your review. As described above, I have also made modifications

dayatsin-amd replied:
I added a patch to the series to reduce the number of #if defined(__linux__). 
Please review. Otherwise lgtm.

hongkzha-amd replied:
Thank you for your patch. I have updata some of the IsDxg check logic.

hongkzha-amd committed:
rocr: Skip CPU mapping operations on WSL

On WSL, CPU cannot access GPU VRAM due to platform restrictions.
CPU access would fault-in system RAM instead, causing data corruption
and memory leaks. Return HSA_STATUS_ERROR to fail fast rather than
silently creating broken mappings. GPU-to-GPU mappings remain functional.

Signed-off-by: Flora Cui <flora.cui@amd.com>
Signed-off-by: Horatio Zhang <Hongkun.Zhang@amd.com>

dayatsin-amd committed:
rocr: reduce ifdef linux

v2: Fix IsDXG check logic

Signed-off-by: David Yat Sin <David.YatSin@amd.com>
Signed-off-by: Horatio Zhang <Hongkun.Zhang@amd.com>

manocharahul commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/runtime.cpp:
```diff
@@ -3829,12 +3831,11 @@ hsa_status_t Runtime::MappedHandleAllowedAgent::EnableAccess(hsa_access_permissi
-      }
-      if (!rocr::os::MapMemory(va, size, PermissionsToMemProt(perms), mappedHandle->drm_fd,
-                                reinterpret_cast<uint64_t>(mappedHandle->drm_cpu_addr))) {
+      if (!rocr::os::ProtectMemory(va, size, PermissionsToMemProt(perms))) {
```
UnmapMemory was calling VirtualFree to Decommit memory on windows and MapMemory was not doing anything since memory is already decommit. With this ProtectMemory call, we will break that sequence and memory will not be decommited. This not the correct way to handle this in windows.

hongkzha-amd replied:
Thank you very much for your valuable suggestion. I have updated the code for ProtectMemory in os_win.cpp. Could you please help me review it

manocharahul commented on projects/rocr-runtime/runtime/hsa-runtime/core/util/win/os_win.cpp:
```diff
@@ -470,6 +470,11 @@ bool MapMemory(void* addr, size_t size, MemProt perms, int fd [[maybe_unused]],
 
+bool ProtectMemory(void* va, size_t size, MemProt perms) {
+  DWORD oldProt;
+  return VirtualProtect(va, size, memProtToOsProt(perms), &oldProt) != 0;
```
add this line before call to VirtualProtect.

`if (perms == MEM_PROT_NONE) {
     UncommitMemory(addr, size);
     return
}`

hongkzha-amd replied:
Got it! Thanks for your suggestion. I have made the corresponding modifications

dayatsin-amd commented:
Changes look ok on my side.
@manocharahul please also approve if you are ok.

hongkzha-amd committed:
Merge branch 'develop' into users/hongkzha/drop-drm-fd-mmap-in-wsl

hongkzha-amd committed:
Merge branch 'develop' into users/hongkzha/drop-drm-fd-mmap-in-wsl
